assets/js/frontend/cart.js

```diff
@@ -17,24 +53,139 @@ jQuery( function( $ ) {
 			shipping_methods[ $( this ).data( 'index' ) ] = $( this ).val();
 		});
 
-		$( 'div.cart_totals' ).block({
-			message: null,
-			overlayCSS: {
-				background: '#fff',
-				opacity: 0.6
-			}
-		});
+		block( $( 'div.cart_totals' ) );
 
 		var data = {
 			security: wc_cart_params.update_shipping_method_nonce,
 			shipping_method: shipping_methods
 		};
 
-		$.post( wc_cart_params.wc_ajax_url.toString().replace( '%%endpoint%%', 'update_shipping_method' ), data, function( response ) {
+		$.post( get_url( 'update_shipping_method' ), data, function( response ) {
 			$( 'div.cart_totals' ).replaceWith( response );
 			$( document.body ).trigger( 'updated_shipping_method' );
 		});
 	});
 
 	$( '.shipping-calculator-form' ).hide();
+
+	// Update the cart after something has changed.
+	var update_cart_totals = function() {
+		block( $( 'div.cart_totals' ) );
+
+		$.ajax( {
+			url:      get_url( 'get_cart_totals' ),
+			dataType: 'html',
+			success: function( response ) {
+				$( 'div.cart_totals' ).replaceWith( response );
+			}
+		});
+	};
+
+	// clears previous notices and shows new one above form.
+	var show_notice = function( html_element ) {
+		var $form = $( 'div.woocommerce > form' );
+
+		$( '.woocommerce-error, .woocommerce-message' ).remove();
+		$form.before( html_element );
+	};
+
+	// Coupon code
+	$( '[name="apply_coupon"]' ).on( 'click', function( evt ) {
+		evt.preventDefault();
+
+		var $form = $( 'div.woocommerce > form' );
+
+		if ( is_blocked( $form ) ) {
+			return false;
+		}
+
+		block( $form );
+
+		var $text_field = $( '#coupon_code' );
+		var coupon_code = $text_field.val();
+
+		var data = {
+			security: wc_cart_params.apply_coupon_nonce,
+			coupon_code: coupon_code
+		};
+
+		$.ajax( {
+			type:     'POST',
+			url:      get_url( 'apply_coupon' ),
+			data:     data,
+			dataType: 'html',
+			success: function( response ) {
+				show_notice( response );
+			},
+			complete: function() {
+				unblock( $form );
+				$text_field.val( '' );
+				update_cart_totals();
+			}
+		});
+	});
+
+	$( document ).on( 'click', 'a.woocommerce-remove-coupon', function( evt ) {
+		evt.preventDefault();
+
+		var $tr = $( this ).parents( 'tr' );
+		var coupon = $( this ).attr( 'data-coupon' );
+
+		block( $tr.parents( 'table' ) );
+
+		var data = {
+			security: wc_cart_params.remove_coupon_nonce,
+			coupon: coupon
+		};
+
+		$.ajax( {
+			type:    'POST',
+			url:      get_url( 'remove_coupon' ),
+			data:     data,
+			dataType: 'html',
+			success: function( response ) {
+				show_notice( response );
+				unblock( $tr.parents( 'table' ) );
+			},
+			complete: function() {
+				update_cart_totals();
+			}
+		});
```

Comment:
I generally just focus on updating/fixing them when the lines are already changing (instead of making extra changes for the sake of coding standards).